### PR TITLE
Refactor to remove usage of post related effects in packages/editor.

### PR DESCRIPTION
**Pull request by nerrad**
## Description

This pull is the first step in moving away from the lingering usage of effects in various data stores among packages.  This pull specifically will deal with the `@wordpress/editor` package (`core/editor` store).

For the most part the following goals are in mind:

- organize file structure so things are a bit more clearly sorted.
- As much as possible I'm preserving the existing logic.  This means shifting up an action creator that returns a action.type that had effects into an action generator (i.e. `savePost`) and then moving the effects into the action-generator. Dispatches and selects in the effects would then be moved into the generator and yield _control_ dispatch and selects (same for usage apiFetch etc).
- Of course refactoring tests for all the changes.
- This will also allow me to keep things in the existing files until the last moment  when I'm ready to roll over to using the new code and test.

## How has this been tested?

* [x] covered changes in unit tests
* [x] existing e2e tests coverage
* [x] Tested various post saves and status changes via browser testing to ensure there's no issues.

## Types of changes

There are no breaking changes in this pull.  The interfaces remain effectively the same. However:

- See the note below regarding the trashPost action and some consistency introduction there. The signature has changed for `trashPost` but ultimately that does not affect any behaviour.
- Some new actions are exposed to retain existing behaviour from the effects.  These actions are not really something that should be publicly exposed so they are prefixed with `__experimental` so that provides room for later iterations potentially changing/removing them.  I didn't do any refactoring there because the goal of this pull is to just convert from effects to controls/action-generators.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 

### `trashPost` action.

Currently `trashPost` accepts a `postId` and a `postTypeSlug` but they are inherently ignored by the effect callback handling the actual trash in favor of trashing the currentPost.  I [brought this up in slack](https://wordpress.slack.com/archives/C02QB2JS7/p1549590206300100) and @aduth and I resolved that:

- `core/editor` can keep `trashPost` but drop the args.  The interface then assumes that `trashPost` means trash the current post in the state for the editor.
- at some point in the future we could introduce a more explicit `trashPost( postId, postType )` for the `core` store that would trash the given post for the given id.

Related #11124

### CODE REVIEW

**Comment by nerrad** [comment]
Need to implement `createRegistryControl` from #13722 when it lands in master

**Comment by nerrad** [comment]
Effects to be refactored:

* [ ] REQUEST_POST_UPDATE
* [ ] REQUEST_POST_UPDATE_SUCCESS
* [ ] REQUEST_POST_UPDATE_FAILURE
* [ ] TRASH_POST
* [ ] TRASH_POST_FAILURE
* [ ] REFRESH_POST
* [ ] MERGE_BLOCKS
* [ ] SETUP_EDITOR
* [ ] RESET_BLOCKS
* [ ] SYNCHRONIZE_TEMPLATE
* [ ] FETCH_REUSABLE_BLOCKS
* [ ] SAVE_REUSABLE_BLOCK
* [ ] DELETE_REUSABLE_BLOCK
* [ ] RECEIVE_REUSABLE_BLOCKS
* [ ] CONVERT_BLOCK_TO_STATIC
* [ ] CONVERT_BLOCK_TO_REUSABLE
* [ ] REMOVE_BLOCKS
* [ ] REPLACE_BLOCKS
* [ ] MULTI_SELECT

**Comment by youknowriad** [comment]
Do you think it's possible to ship PRs for these effects one after the other?

**Comment by nerrad** [comment]
I'm wary about doing multiple pulls because I'm not just refactoring I'm also re-organizing.  That's a lot of potential for merge conflicts.  With that said, what I could do is ship all the post related effect refactors first and then do the block related affects in a different pull.  Would something like that work better?

**Comment by youknowriad** [comment]
> Would something like that work better?

Yes, If it's a big PR, it's very hard to review properly. splitting is better.

**Comment by nerrad** [comment]
Getting close to done.  However there are some failing e2e tests I need to look at (initial look shows it appears to be something involving autopost saves).

Browser testing has gone okay for me so far.  If anyone wants to start reviewing this that would be fine.  I'll keep looking into why the tests are failing.

**Comment by nerrad** [comment]
So browser testing is working great for me.  I _think_ the issue with the e2e tests is caused by `post.id` not set in the `savePost` action triggered by autosave in the e2e tests.  It's possible there's some special autosave handling in the e2e tests that is affected by this refactor (haven't dug in yet).  Just making these notes for myself so I don't forget.

**Comment by nerrad** [comment]
> So browser testing is working great for me.

Sigh, I didn't have the GB plugin active :/  I'm reproducing issues while user testing so clearly something is borked.

**Comment by nerrad** [comment]
whoah, found the cause.  The issue is `redux-routine` determines if something "quacks like an action" by the presence of a `type` property.  The selector for `getCurrentPost` is return a post object which has a type property.  So `redux-routine` considers that another action to dispatch :/

**Comment by aduth** [comment]
> whoah, found the cause. The issue is `redux-routine` determines if something "quacks like an action" by the presence of a `type` property. The selector for `getCurrentPost` is return a post object which has a type property. So `redux-routine` considers that another action to dispatch :/

Related Slack conversation ([link requires registration](https://make.wordpress.org/chat/)):

https://wordpress.slack.com/archives/C02QB2JS7/p1549904073023400

**Comment by nerrad** [comment]
I've tested the change in https://github.com/WordPress/gutenberg/pull/13818 in this branch and it resolves the issues I was seeing here.  So once that pull lands and we get it in this branch this should be good to go to (after review of course)

**Comment by nerrad** [comment]
rats. still some e2e fails (and linting).  I'm currently debugging the e2e fails locally.

**Comment by nerrad** [comment]
Okay there is only one remaining failing test and it is:

> The command "npm run check-local-changes" exited with 1.

I'm not sure how to resolve this.  I imagine it has to do with a package-lock change when travis runs npm build.  So does a package-lock change need to be generated in a commit? Should it happen in a specific environment?

File: packages/editor/src/store/actions/index.js
Comment (aduth) [inline packages/editor/src/store/actions/index.js]: Is there a compelling reason to segregate the two?

Comment (nerrad) [reply packages/editor/src/store/actions/index.js]: Do you mean among the post actions?  or do you mean specifically having "post" type actions and eventually "block" type actions?

If the former, I find it easier to know where to go for general actions as opposed to action generators.  It also aids with splitting up tests to cover the specific things.

Comment (aduth) [reply packages/editor/src/store/actions/index.js]: > I find it easier to know where to go for general actions as opposed to action generators.

It doesn't seem to me that it's something someone should need to distinguish. I think a current reality is that we have spaghetti generators, but they're all just "actions (dispatchers)", ideally returning or yielding simple action objects.

Comment (nerrad) [reply packages/editor/src/store/actions/index.js]: I'm probably ignorant of potential reasons why we wouldn't want to split up things in this case.  Are there technical/performance/build reasons why its good to avoid splitting up files like this?

Comment (nerrad) [reply packages/editor/src/store/actions/index.js]: I usually try to logically split files up into smaller components when I can because:

a. it helps with finding things quicker when I'm just looking for a "class" of logic
b. it makes for smaller number of lines in a file, so less scrolling.
c. My IDE built-in linting etc doesn't get bogged down as much due to larger files.

But as I already mentioned, I'm aware I may be ignorant of good reasons for not splitting things up as I did here.

Comment (aduth) [reply packages/editor/src/store/actions/index.js]: To me, it's not a technical point, it's of discoverability and maintainability generally. What we're doing here is creating a subclass (of actions) where otherwise we've aligned to a common pattern of identifying a few top-level behaviors associated with a store: mutations (`actions.js`), state (`reducer.js`), derivations / getters (`selectors.js`), and side-effects (`effects.js`).

I'm not saying this can't be expanded, but we should do so carefully, deliberately, and consistently. Small files are not inherently more maintainable than a single large file, if it's not obvious or predictable why the files are split the way they are. I find the `effects/` folder pattern to be a good example of this, where I think it was better off in single file form, because the current split is not obvious, and in the majority of cases I arrive at the wrong place when trying to track down effects-related behavior.

My prior point was that the division between generator-actions and non-generator-actions is a bit weak, and arguably intentionally so (my point about "you shouldn't need to distinguish"), so creating subclasses within actions for these does not seem an obvious thing to do.

Comment (nerrad) [reply packages/editor/src/store/actions/index.js]: Kinda funny that we're both hinging our argument on discoverability and maintainability :).  I'm not a project maintainer so I don't want to break any established conventions for the project.  I'm happy to just combine them into one file (and the same with tests) if desired.  So just for clarification, is this a _directive_ I need to follow in order to get this pull mergeable or is this a preference you are arguing for but is not a directive yet?

To be clear, I did loosely swing my intentions by @youknowriad a bit in dm with him before embarking, although he suggested this structure (which mirrors the current effects):

> Personally, I'd start with adding those to `actions.js` but maybe split it by domains
>
> /actions/index.js
> /actions/post.js
> /actions/reusable-blocks.js

So I _did_ expand on that a bit in my implementation.  So in your clarification, can you also give direction for whether I should just put everything into the existing `actions.js` file (and subsequently add all the tests to the existing `tests/actions.js` as well?

Comment (aduth) [reply packages/editor/src/store/actions/index.js]: Nope, no directives. Just surfacing my own opinions stemming from the initial concern / probe for rationale of the split. It seems there's at least a second opinion in your favor from @youknowriad , but I'd go on record to disagree with the "domain"-based split (it wasn't but an hour after my previous complaint about the effects structuring that I wrongly navigated to `effects/posts.js` as my first destination in pursuit of [this snippet of code](https://github.com/WordPress/gutenberg/blob/a99c0b5487ff3639c05a7ce1dbf7823b821382f8/packages/editor/src/store/effects.js#L206-L216)).

To your question, yes, I'd see it all just being in a single `actions.js` file.

There's an interesting overlap between this discussion and some older ones like in #3030, where I think ultimately it was the introduction of `@wordpress/data` and its store-namespacing that served as the alternative to this problem of files becoming large and unwieldy, where the separation was by _stores_ (modules), not by separate files within a store.

Comment (youknowriad) [reply packages/editor/src/store/actions/index.js]: `actions.js` is better if we had one file but if the file grows indefinitely, it becomes hard to navigate the file. maybe there's a middleground to find at some point.

I also said to @nerrad that in general I'm not very opinionated on these matters :)

Comment (nerrad) [reply packages/editor/src/store/actions/index.js]: > but if the file grows indefinitely, it becomes hard to navigate the file.

Ya this is one criteria I use when deciding whether to split up a file.  For example _just_ with the post actions here, combined, they would be 487 lines.  That's _before_ adding in all the block and editor related actions.

> as my first destination in pursuit of [this snippet of code](https://github.com/WordPress/gutenberg/blob/a99c0b5487ff3639c05a7ce1dbf7823b821382f8/packages/editor/src/store/effects.js#L206-L216)).

@aduth my _plan_ was to have `editor-actions` and `block-actions` files that would accompany the `post-actions`.  That snippet is found within the `SETUP_EDITOR` editor block. 

Would a middle ground here be to not split between action generators and action functions and just have `post-actions`, `editor-actions`, and `block-actions`?  That might also be useful for aiding future refactors where some (or all) of the block specific actions are moved elsewhere.

Comment (nerrad) [reply packages/editor/src/store/actions/index.js]: Keep in mind, I'm not only considering the line count for a combined actions file, but also the line count that would be exponentially larger for the corresponding tests file.  Currently combining `actions/test/post.js` and `actions/test/post-generators.js` into one file results in 759 lines.  As soon as a file gets over roughly 500 lines I find it starts to get unwieldy.  I imagine the line count for a combined actions file of editor, post, and block actions will start to go over 2000 easily.

Comment (nerrad) [reply packages/editor/src/store/actions/index.js]: > There's an interesting overlap between this discussion and some older ones like in #3030, where I think ultimately it was the introduction of `@wordpress/data` and its store-namespacing that served as the alternative to this problem of files becoming large and unwieldy, where the separation was by _stores_ (modules), not by separate files within a store.

To explore this further then, are you suggesting that we might want to consider applying some more granular namespacing here and split these up into more stores?  Of course, that could introduce some complications with deprecating the old actions etc but might be preferable over the file splitting.  So we could end up with something like `core/editor/posts`, `core/editor` (for general editor actions), and `core/editor/blocks` for block actions?

File: packages/editor/src/store/controls.js
Comment (nerrad) [inline packages/editor/src/store/controls.js]: It'd be nice if the registry exposed some sort of helper for determining if a selector has a resolver (i.e. `registry.getStore( reducerKey ).hasResolverFor( selectorName )`).  That way this code could be condensed to just having a `select` control that conditionally resolves if there is a resolver.

Comment (youknowriad) [reply packages/editor/src/store/controls.js]: Agreed, that some help from the registry here would be good, this could be addressed in a separate PR maybe?

**Comment by nerrad** [comment]
Okay this branch is up to date and I've refactored the work in here so all the post actions are condensed into `actions.js` (tests updated etc).  Tests pass locally, should be ready for review.  I am a bit surprised that functions prefixed with `__experimental` are showing up in the generated docs.  Not sure if that's something needing addressed or not.

**Comment by aduth** [comment]
> Not sure if that's something needing addressed or not.

It does, but not here.

File: packages/editor/src/store/actions.js
Comment (aduth) [inline packages/editor/src/store/actions.js]: From which guideline are you asserting the documentation standard proposed here?

Elsewhere, the `@param` and `@return` are treated as separately grouped, within each group aligned on the name and descriptions, as documented here:

https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions

There are multiple instances elsewhere in the pull request.

File: packages/editor/src/store/actions.js
Comment (aduth) [inline packages/editor/src/store/actions.js]: There's only one parameter to this function: an object. If the properties of the object are to be documented, they should be as prescribed by JSDoc and in the JavaScript Documentation Standards, e.g. `options.previousPost`

http://usejsdoc.org/tags-param.html#parameters-with-properties
https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions

There are multiple instances elsewhere in the pull request.

File: packages/editor/src/store/actions.js
Comment (aduth) [inline packages/editor/src/store/actions.js]: I'm curious whether we should bother with this constant, considering there is precedent via hundreds of other components defined in the editor module referencing `'core/editor'` directly.

File: packages/editor/src/store/actions.js
Comment (aduth) [inline packages/editor/src/store/actions.js]: Could this be simplified to just `yield* savePost( { /* ... */ } )j` ?

File: packages/editor/src/store/actions.js
Comment (aduth) [inline packages/editor/src/store/actions.js]: I understand how you're using this for the unit tests, but it does not seem sensible to me that a default export from this file be a subset of generator actions.

Would it be possible instead to consider for the purposes of `jest.spyOn`, to use the `import * as actions from '../actions';` syntax?

File: packages/editor/src/store/constants.js
Comment (aduth) [inline packages/editor/src/store/constants.js]: Where is this referenced? Was there an intention to consolidate constants from the `selectors.js` file here? Because that does not appear to be represented (the constants remain in `selectors.js` and are referenced as defined there).

File: packages/editor/src/store/controls.js
Comment (aduth) [inline packages/editor/src/store/controls.js]: This could be simplified to:

```js
import triggerFetch from '@wordpress/api-fetch';
```

File: packages/editor/src/store/controls.js
Comment (aduth) [inline packages/editor/src/store/controls.js]: We had a perfectly good DocBlock which was abolished?

File: packages/editor/src/store/controls.js
Comment (aduth) [inline packages/editor/src/store/controls.js]: Minor: We could optimize this to avoid calling `getResult` until after we're certain it would be useful for the satisfied result of `hasFinished()`, i.e. call it inline with `return resolve( getResult() );` as done in the subscriber callback, or at least assigned as part of the `if` condition body.

File: packages/editor/src/store/utils/notice-builder.js
Comment (aduth) [inline packages/editor/src/store/utils/notice-builder.js]: By convention, I'd recommend aligning all but the first lines with the start of the description from the previous line. This isn't documented in the JavaScript Documentation Standards, but is mentioned in passing in the PHP standards:

https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#1-1-parameters-that-are-arrays

>(aligned with Description, if wraps to a new line)

**Review by aduth — COMMENTED** [review]
This seems like a solid refactor.

In case you're feeling particularly averse to merge conflicts, it might benefit you to squash most / all of your commits within the branch.

File: packages/editor/src/store/actions.js
Comment (nerrad) [reply packages/editor/src/store/actions.js]: Huh, I've honestly never noticed this before.  Will correct.

File: packages/editor/src/store/actions.js
Comment (nerrad) [reply packages/editor/src/store/actions.js]: > There's only one parameter to this function: an object.

Sorry I don't follow.  There's multiple arguments on this function not one.

File: packages/editor/src/store/actions.js
Comment (nerrad) [reply packages/editor/src/store/actions.js]: I think having the constant helps prevent against typos personally (however I do understand that reasoning can't be extrapolated to the store/module/reducer key being referenced outside of the package).

Another nit is that I see a variety of terms used to reference this argument in the codebase, sometimes its "reducer key", sometimes its "module key" and sometimes it is "store key".  I just went with `MODULE_KEY` here because that was the variable name given this originally when registering the store.

File: packages/editor/src/store/actions.js
Comment (nerrad) [reply packages/editor/src/store/actions.js]: I used `generatorActions` here so I could mock the `savePost` action more easily.  It's possible I'm overlooking a way to more easily mock this in the tests or the mocking is unnecessary (we could just check that it yields a generator instead).

Comment (nerrad) [reply packages/editor/src/store/actions.js]: saw the subsequent comment about `generatorActions` and realize that ya I can just yield savePost directly I think here.

File: packages/editor/src/store/actions.js
Comment (nerrad) [reply packages/editor/src/store/actions.js]: doh, ya, I'll see if that works sufficiently for tests.

File: packages/editor/src/store/constants.js
Comment (nerrad) [reply packages/editor/src/store/constants.js]: hmm ya, intended to consolidate all constants into one file.  Makes it much easier to reference them (and helps with future refactors).  I'll review/fix.

File: packages/editor/src/store/controls.js
Comment (nerrad) [reply packages/editor/src/store/controls.js]: doh yup.

File: packages/editor/src/store/controls.js
Comment (nerrad) [reply packages/editor/src/store/controls.js]: hmm not intentional, maybe an artifact from resolving a merge conflict maybe.

File: packages/editor/src/store/controls.js
Comment (nerrad) [reply packages/editor/src/store/controls.js]: hmm I just copied this directly from the `resolveSelector` function in `effects/utils` assuming via the comment that the selector needed to be called to trigger the resolution (otherwise `hasFinished()` would never return true).  The comment block made sense to me here.  Still feel it's unnecessary?

Comment (aduth) [reply packages/editor/src/store/controls.js]: Ah, I neglected to consider the preceding comment. It feels awkward to rely on a side-effect from a function named as a getter, but I suppose that's the nature of resolvers. It should be fine as-is.

File: packages/editor/src/store/actions.js
Comment (nerrad) [reply packages/editor/src/store/actions.js]: Oh I follow now.  I think this was the doc block from the original function that I just copied over (but I could be wrong), I'll fix.

File: packages/editor/src/store/actions.js
Comment (nerrad) [reply packages/editor/src/store/actions.js]: Won't work, need to keep this as is so it can be tested (see comment in other thread).

File: packages/editor/src/store/actions.js
Comment (nerrad) [reply packages/editor/src/store/actions.js]: Turns out doing `import * as actions` and then `jest.spyOn( actions, 'savePost')` does not work.

File: packages/editor/src/store/controls.js
Comment (nerrad) [reply packages/editor/src/store/controls.js]: Take note that this doc block isn't really accurate, so I suspect this was a merge conflict artifact (and likely the reason I resolved the way I did).  I suspect the doc block was on some other control action in here so to resolve, I'll just add doc blocks for all the actions :)

File: packages/editor/src/store/actions.js
Comment (nerrad) [reply packages/editor/src/store/actions.js]: So turns out my IDE (phpStorm) doesn't have much ability to customize the jsdocs template it creates.  So I've just never noticed that its defaults diverge so much from the wp core standards (and thus in my own projects I've just never thought anything of it).  I'll try to watch for this in the future.

File: packages/editor/src/store/utils/notice-builder.js
Comment (nerrad) [reply packages/editor/src/store/utils/notice-builder.js]: The [WP standards](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions) you linked to earlier have descriptions aligned _per group_.  So for instance the @params have their descriptions aligned, and the @return is not necessarily aligned with the previous group (at least from the example given).  So that's what I followed in my pass through the doc blocks.

**Comment by nerrad** [comment]
Looks like tests are passing except for the undo related e2e tests that are currently failing everywhere else too.

File: packages/editor/src/store/actions.js
Comment (aduth) [reply packages/editor/src/store/actions.js]: > Another nit is that I see a variety of terms used to reference this argument in the codebase, sometimes its "reducer key", sometimes its "module key" and sometimes it is "store key". I just went with `MODULE_KEY` here because that was the variable name given this originally when registering the store.

Yes, it's been a lingering problem needing to be addressed. There wasn't already an issue, so I created one at #14150.

File: packages/editor/src/store/actions.js
Comment (aduth) [reply packages/editor/src/store/actions.js]: > Turns out doing `import * as actions` and then `jest.spyOn( actions, 'savePost')` does not work.

Why not?

File: packages/editor/src/store/utils/notice-builder.js
Comment (aduth) [reply packages/editor/src/store/utils/notice-builder.js]: > The [WP standards](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/#functions) you linked to earlier have descriptions aligned _per group_. So for instance the @params have their descriptions aligned, and the @return is not necessarily aligned with the previous group (at least from the example given)

Yes, this sounds right. My point was toward the `@return` specifically aligning with itself. In your revisions, I think there are issues with mixed whitespace / tabbing.

image

File: packages/editor/src/store/actions.js
Comment (nerrad) [reply packages/editor/src/store/actions.js]: No idea.  I tried it:

```js
import * as actions from '../actions';

// .... later in the test
describe( 'autosave()', () => {
		let savePostSpy;
		beforeAll( () => savePostSpy = jest.spyOn( actions, 'savePost' ) );
		afterAll( () => savePostSpy.mockRestore() );
		// autosave is mostly covered by `savePost` tests so just test the correct call
		it( 'calls savePost with the correct arguments', () => {
			const fulfillment = actions.autosave();
			fulfillment.next();
			expect( savePostSpy ).toHaveBeenCalled();
			expect( savePostSpy ).toHaveBeenCalledWith( { isAutosave: true } );
		} );
	} );
```

The test fails with the mock not getting called.  When I use `generatorActions` the mock works as expected.  I don't doubt there may be something I'm doing wrong but I'm guessing its because the mock isn't actually getting invoked by the autosave call.

File: packages/editor/src/store/utils/notice-builder.js
Comment (nerrad) [reply packages/editor/src/store/utils/notice-builder.js]: argh, looks fine in the IDE.  I *hate* when the IDE doesn't cooperate with documentation requirements.  I'll revisit but likely won't get to it today.

File: packages/editor/src/store/actions.js
Comment (aduth) [reply packages/editor/src/store/actions.js]: I sense it should be possible. The issue is likely that the spy creates a new reference which is different than the one called internal to the underlying `actions.js` file.

But I also wonder: Maybe rather than testing that it was called, we could test that the result of calling autosave would be the same as the result of calling `savePost` with the expected arguments?

Something like:

```js
it( 'calls savePost with the correct arguments', () => {
	expect( Array.from( autosave() ) ).toEqual( Array.from( savePost( { isAutosave: true } ) ) );
} );
```

Comment (nerrad) [reply packages/editor/src/store/actions.js]: Not sure how practical/useful that will be because they are both generators.

Comment (nerrad) [reply packages/editor/src/store/actions.js]: So one thing I _could_ do here is instead of yielding the `savePost` generator from `autosave`, I can just yield the `dispatch` control.  That way it's more easily testable and I can get rid of the `generatorActions` default export.

**Comment by nerrad** [comment]
The [latest commits](https://github.com/WordPress/gutenberg/pull/13716/files/6d74643bcfea22604b41d797ad5b221a7db4c11b..4cc4015b16e21d896a492c954f29aefcf31db659)

- fix spacing issues in jsdoc blocks
- switch to use `STORE_KEY` instead of `MODULE_KEY` for the store name reference to contribute to the general renaming schema as initiated by #14150 
- switched `autosave()` generator action to use the dispatch control for dispatching the `savePost` action.  Thus I can remove the `generatorActions` default export and test that the autosave() is dispatching with the right args.

**Comment by aduth** [comment]
> * switched `autosave()` generator action to use the dispatch control for dispatching the `savePost` action.  Thus I can remove the `generatorActions` default export and test that the autosave() is dispatching with the right args.

While normally I'd be reluctant to alter the implementation for the sake of tests (in part the reason for the original review comment), it doesn't really strike me as any better or worse than the original implementation of the action creator to use the `dispatch` control this way. 👍

File: packages/editor/src/store/selectors.js
Comment (aduth) [inline packages/editor/src/store/selectors.js]: I guess these were overlooked in the split for block editor? I verified they're not used in editor.

**Comment by nerrad** [comment]
Ya I agree wrt to reluctance about altering implementation for the sake of tests.   But the benefit of this approach is that there isn't anything _additional_ added for the sake of easier testing (i.e. the previous behaviour).  This is a straight implementation that just happens to be more testable.

**Comment by aduth** [comment]
The notice generated for a failed trash when offline is a bit unexpected. But the same exists on master, so ought be addressed separately. There might already be an issue.

image

**Review by aduth — APPROVED** [review]
Nice work here 👍

File: packages/editor/src/store/selectors.js
Comment (nerrad) [reply packages/editor/src/store/selectors.js]: Either that or I resolved the merge incorrectly when bringing into this branch.

**Pull request merged**
